wallet/src/de/schildbach/wallet/service/CoinJoinService.kt

```diff
@@ -162,61 +156,128 @@ class CoinJoinMixingService @Inject constructor(
                 updateBlockChain(blockChain)
             }
             .launchIn(coroutineScope)
+
+        blockchainStateProvider.observeState()
+            .filterNotNull()
+            .distinctUntilChanged()
+            .onEach { blockChainState ->
+                log.info("coinjoin: new block: ${blockChainState.bestChainHeight}")
+                updateBalance(walletDataProvider.getWalletBalance())
+            }
+            .launchIn(coroutineScope)
+
+        walletDataProvider.observeBalance()
+            .distinctUntilChanged()
+            .onEach { balance ->
+                // switch to our context
+                coroutineScope.launch {
+                    updateBalance(balance)
+                }
+            }
+            .launchIn(uiCoroutineScope) // required for observeBalance
```

Comment:
This might be an oversight in `WalletBalanceObserver`. We should be able to observe balance in any thread. I'll check it when have a chance.